File: backend/team6/src/test/java/programmers/team6/domain/vacation/service/VacationServiceTests.java
package programmers.team6.domain.vacation.service;

import static org.assertj.core.api.Assertions.*;
import static org.mockito.Mockito.*;

import java.time.Duration;
import java.time.LocalDateTime;
import java.util.Optional;

import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.springframework.context.annotation.Import;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.test.context.junit.jupiter.SpringExtension;

import programmers.team6.domain.member.entity.Member;
import programmers.team6.domain.admin.repository.CodeRepository;
import programmers.team6.domain.member.repository.MemberRepository;
import programmers.team6.domain.vacation.dto.request.VacationCreateRequestDto;
import programmers.team6.domain.vacation.dto.response.VacationListResponseDto;
import programmers.team6.domain.vacation.repository.ApprovalStepRepository;
import programmers.team6.domain.vacation.repository.VacationInfoRepository;
import programmers.team6.domain.vacation.repository.VacationRequestRepository;
import programmers.team6.domain.vacation.repository.VacationRequestSearchRepository;
import programmers.team6.domain.vacation.util.mapper.VacationMapper;
import programmers.team6.global.exception.customException.BadRequestException;
import programmers.team6.global.exception.customException.NotFoundException;
import programmers.team6.support.MemberMother;

@ExtendWith(SpringExtension.class)
@Import(VacationMapper.class)
class VacationServiceTests {
	@Mock
	private VacationMapper vacationMapper;
	@Mock
	private VacationInfoRepository vacationInfoRepository;
	@Mock
	private VacationRequestRepository vacationRequestRepository;
	@Mock
	private ApprovalStepRepository approvalStepRepository;
	@Mock
	private MemberRepository memberRepository;
	@Mock
	private CodeRepository codeRepository;
	@Mock
	private ApprovalStepService approvalStepService;
	@Mock
	private VacationRequestSearchRepository vacationRequestSearchRepository;
	@InjectMocks
	private VacationService vacationService;

	@Nested
	class getMyVacationInfo {
		@Test
		void should_throwRuntimeException_when_givenNotExistMemberId() {
			// when
			when(memberRepository.findById(0L)).thenReturn(Optional.empty());

			// then
			assertThatThrownBy(() -> vacationService.getMyVacationInfo(0L)).isInstanceOf(RuntimeException.class);
		}
	}

	@Nested
	class requestVacation {
		@Test
		void should_throwRuntimeException_when_givenNotExistVacationInfo() {
			// when
			when(memberRepository.findByIdWithDeptAndLeader(0L)).thenReturn(Optional.empty());

			// then
			assertThatThrownBy(() -> vacationService.requestVacation(0L, null)).isInstanceOf(RuntimeException.class);
		}

		@Test
		void should_throwBadRequestException_when_givenInvalidFromAndTo() {
			// given
			Member member = MemberMother.member();
			LocalDateTime from = LocalDateTime.now().plusDays(1);
			LocalDateTime to = LocalDateTime.now();
			VacationCreateRequestDto vacationCreateRequestDto = new VacationCreateRequestDto(from, to, "empty",
				"empty");

			// when
			when(memberRepository.findByIdWithDeptAndLeader(0L)).thenReturn(Optional.of(member));
			when(vacationRequestRepository.countInRangeFromBetweenToBy(member.getId(),
				from, to)).thenReturn(1l);

			// then
			assertThatThrownBy(
				() -> vacationService.requestVacation(member.getId(), vacationCreateRequestDto)).isInstanceOf(
				BadRequestException.class);
		}

		@Test
		void should_throwNotFoundException_when_givenInvalidMemberIdAndVacationInfoType() {
			// given
			Member member = MemberMother.member();
			LocalDateTime from = LocalDateTime.now().plusDays(1);
			LocalDateTime to = LocalDateTime.now();
			String vacationType = "vacationType";
			VacationCreateRequestDto vacationCreateRequestDto = new VacationCreateRequestDto(from, to, "reason",
				vacationType);

			// when
			when(memberRepository.findByIdWithDeptAndLeader(0L)).thenReturn(Optional.of(member));
			when(vacationRequestRepository.countInRangeFromBetweenToBy(member.getId(),
				vacationCreateRequestDto.getFrom(), vacationCreateRequestDto.getTo())).thenReturn(0l);
			when(vacationRequestRepository.calculateRequestedVacationDays(from, to, vacationType)).thenReturn(0d);
			when(vacationInfoRepository.findActualRemainingVacationDays(member.getId(), vacationType)).thenReturn(
				Optional.empty());

			// then
			assertThatThrownBy(
				() -> vacationService.requestVacation(member.getId(), vacationCreateRequestDto)).isInstanceOf(
				NotFoundException.class);
		}

		@Test
		void should_throwBadRequestException_when_actualRemainCountLessThanRequestDays() {
			// given
			Member member = MemberMother.member();
			LocalDateTime from = LocalDateTime.now().plusDays(1);
			LocalDateTime to = LocalDateTime.now();
			String vacationType = "vacationType";
			VacationCreateRequestDto vacationCreateRequestDto = new VacationCreateRequestDto(from, to, "reason",
				vacationType);

			// when
			when(memberRepository.findByIdWithDeptAndLeader(0L)).thenReturn(Optional.of(member));
			when(vacationRequestRepository.countInRangeFromBetweenToBy(member.getId(),
				vacationCreateRequestDto.getFrom(), vacationCreateRequestDto.getTo())).thenReturn(0l);
			when(vacationRequestRepository.calculateRequestedVacationDays(from, to, vacationType)).thenReturn(5d);
			when(vacationInfoRepository.findActualRemainingVacationDays(member.getId(), vacationType)).thenReturn(
				Optional.of(0d));

			// then
			assertThatThrownBy(
				() -> vacationService.requestVacation(member.getId(), vacationCreateRequestDto)).isInstanceOf(
				BadRequestException.class);
		}

		@Test
		void should_throwRuntimeException_when_givenInvalidGroupCodeAndCode() {
			// given
			Member member = MemberMother.member();
			Member leader = MemberMother.member();
			member.getDept().appointLeader(leader);

			LocalDateTime from = LocalDateTime.now().plusDays(1);
			LocalDateTime to = LocalDateTime.now();
			String vacationType = "vacationType";
			VacationCreateRequestDto vacationCreateRequestDto = new VacationCreateRequestDto(from, to, "reason",
				vacationType);

			// when
			when(memberRepository.findByIdWithDeptAndLeader(0L)).thenReturn(Optional.of(member));
			when(vacationRequestRepository.countInRangeFromBetweenToBy(member.getId(),
				vacationCreateRequestDto.getFrom(), vacationCreateRequestDto.getTo())).thenReturn(0l);
			when(vacationRequestRepository.calculateRequestedVacationDays(from, to, vacationType)).thenReturn(0d);
			when(vacationInfoRepository.findActualRemainingVacationDays(member.getId(), vacationType)).thenReturn(
				Optional.of(1d));

			// then
			assertThatThrownBy(
				() -> vacationService.requestVacation(member.getId(), vacationCreateRequestDto)).isInstanceOf(
				RuntimeException.class);
		}

	}

	@Test
	void should_saveVacation_when_givenValidMemberIdAndVacationCreateRequestDto() {
		// given
		Member member = MemberMother.member();
		Member leader = MemberMother.member();
		member.getDept().appointLeader(leader);

		LocalDateTime from = LocalDateTime.now();
		LocalDateTime to = from.plusDays(2);
		VacationCreateRequestDto vacationCreateRequestDto = new VacationCreateRequestDto(from, to, "reason",
			"vacationType");

		// when
		when(memberRepository.findByIdWithDeptAndLeader(member.getId())).thenReturn(Optional.of(member));
		when(vacationRequestRepository.countInRangeFromBetweenToBy(member.getId(), from, to)).thenReturn(0l);
		when(vacationRequestRepository.calculateRequestedVacationDays(from, to,
			vacationCreateRequestDto.getVacationType())).thenReturn((double)Duration.between(from, to).toDays());
		when(vacationInfoRepository.findActualRemainingVacationDays(member.getId(),
			vacationCreateRequestDto.getVacationType())).thenReturn(Optional.of(30d));
		when(codeRepository.findByGroupCodeAndCode("VACATION_TYPE",
			vacationCreateRequestDto.getVacationType())).thenReturn(Optional.empty());

		// then
		assertThatThrownBy(
			() -> vacationService.requestVacation(member.getId(), vacationCreateRequestDto)).isInstanceOf(
			RuntimeException.class);
	}

	@Nested
	class GetVacationRequestList {

		@Test
		void should_returnEmptyResult_when_findIdPageIsEmpty() {
			// given
			Member member = MemberMother.member();

			// when
			when(memberRepository.findById(member.getId())).thenReturn(Optional.of(member));
			when(vacationRequestRepository.findIdsByRequesterIdPaging(anyLong(), any(Pageable.class))).thenReturn(
				Page.empty());

			// then
			VacationListResponseDto result = vacationService.getVacationRequestList(member.getId(), 0);
			assertThat(result.getContent()).isEmpty();
			assertThat(result.getPageNumber()).isEqualTo(0);
		}
	}

	@Nested
	class UpdateVacationRequest {
		/**
		 * requestVacation 메서드와 겹치는게 많아서 리팩토링 후 테스트 추가가 필요해 보임
		 * @author gunwoong
		 */
	}

}

File: backend/team6/src/test/java/programmers/team6/domain/vacation/service/VacationServiceTests.kt
package programmers.team6.domain.vacation.service

import io.mockk.every
import io.mockk.mockk
import org.assertj.core.api.Assertions.assertThatThrownBy
import org.junit.jupiter.api.Nested
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.springframework.context.annotation.Import
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.data.repository.findByIdOrNull
import org.springframework.test.context.junit.jupiter.SpringExtension
import programmers.team6.domain.admin.repository.CodeRepository
import programmers.team6.domain.member.repository.MemberRepository
import programmers.team6.domain.vacation.dto.request.VacationCreateRequestDto
import programmers.team6.domain.vacation.repository.ApprovalStepRepository
import programmers.team6.domain.vacation.repository.VacationInfoRepository
import programmers.team6.domain.vacation.repository.VacationRequestRepository
import programmers.team6.domain.vacation.repository.VacationRequestSearchRepository
import programmers.team6.domain.vacation.util.mapper.VacationMapper
import programmers.team6.global.exception.customException.BadRequestException
import programmers.team6.global.exception.customException.NotFoundException
import programmers.team6.support.MemberMother
import java.time.Duration
import java.time.LocalDateTime
import java.util.*

@ExtendWith(SpringExtension::class)
@Import(VacationMapper::class)
class VacationServiceTests {

    private val vacationMapper: VacationMapper = mockk()
    private val vacationInfoRepository: VacationInfoRepository = mockk()
    private val vacationRequestRepository: VacationRequestRepository = mockk()
    private val approvalStepRepository: ApprovalStepRepository = mockk()
    private val memberRepository: MemberRepository = mockk()
    private val codeRepository: CodeRepository = mockk()
    private val approvalStepService: ApprovalStepService = mockk()
    private val vacationRequestSearchRepository: VacationRequestSearchRepository = mockk()

    private val vacationService = VacationService(
        vacationInfoRepository,
        vacationMapper,
        vacationRequestRepository,
        approvalStepRepository,
        memberRepository,
        codeRepository,
        approvalStepService,
        vacationRequestSearchRepository
    )

    @Nested
    inner class GetMyVacationInfo {
        @Test
        fun `should throw RuntimeException when given not exist member id`() {
            // when
            every { memberRepository.findById(0L) } returns Optional.empty()

            // then
            assertThatThrownBy { vacationService.getMyVacationInfo(0L) }
                .isInstanceOf(RuntimeException::class.java)
        }
    }

    @Nested
    inner class RequestVacation {
        @Test
        fun `should throw RuntimeException when given not exist vacation info`() {
            // when
            every { memberRepository.findByIdWithDeptAndLeader(0L) } returns null

            // then
            assertThatThrownBy { vacationService.requestVacation(0L, mockk()) }
                .isInstanceOf(RuntimeException::class.java)
        }

        @Test
        fun `should throw BadRequestException when given invalid from and to`() {
            // given
            val member = MemberMother.member()
            val from = LocalDateTime.now().plusDays(1)
            val to = LocalDateTime.now()
            val vacationCreateRequestDto = VacationCreateRequestDto(from, to, "empty", "empty")

            // when
            every { memberRepository.findByIdWithDeptAndLeader(0L) } returns member
            every {
                vacationRequestRepository.countInRangeFromBetweenToBy(member.id!!, from, to)
            } returns 1L

            // then
            assertThatThrownBy {
                vacationService.requestVacation(member.id!!, vacationCreateRequestDto)
            }.isInstanceOf(BadRequestException::class.java)
        }

        @Test
        fun `should throw NotFoundException when given invalid member id and vacation info type`() {
            // given
            val member = MemberMother.member()
            val from = LocalDateTime.now().plusDays(1)
            val to = LocalDateTime.now()
            val vacationType = "vacationType"
            val vacationCreateRequestDto = VacationCreateRequestDto(from, to, "reason", vacationType)

            // when
            every { memberRepository.findByIdWithDeptAndLeader(0L) } returns member
            every {
                vacationRequestRepository.countInRangeFromBetweenToBy(
                    member.id!!,
                    vacationCreateRequestDto.from,
                    vacationCreateRequestDto.to
                )
            } returns 0L
            every {
                vacationRequestRepository.calculateRequestedVacationDays(from, to, vacationType)
            } returns 0.0
            every {
                vacationInfoRepository.findActualRemainingVacationDays(member.id!!, vacationType)
            } returns null

            // then
            assertThatThrownBy {
                vacationService.requestVacation(member.id!!, vacationCreateRequestDto)
            }.isInstanceOf(NotFoundException::class.java)
        }

        @Test
        fun `should throw BadRequestException when actual remain count less than request days`() {
            // given
            val member = MemberMother.member()
            val from = LocalDateTime.now().plusDays(1)
            val to = LocalDateTime.now()
            val vacationType = "vacationType"
            val vacationCreateRequestDto = VacationCreateRequestDto(from, to, "reason", vacationType)

            // when
            every { memberRepository.findByIdWithDeptAndLeader(0L) } returns member
            every {
                vacationRequestRepository.countInRangeFromBetweenToBy(
                    member.id!!,
                    vacationCreateRequestDto.from,
                    vacationCreateRequestDto.to
                )
            } returns 0L
            every {
                vacationRequestRepository.calculateRequestedVacationDays(from, to, vacationType)
            } returns 5.0
            every {
                vacationInfoRepository.findActualRemainingVacationDays(member.id!!, vacationType)
            } returns 0.0

            // then
            assertThatThrownBy {
                vacationService.requestVacation(member.id!!, vacationCreateRequestDto)
            }.isInstanceOf(BadRequestException::class.java)
        }

        @Test
        fun `should throw RuntimeException when given invalid group code and code`() {
            // given
            val member = MemberMother.member()
            val leader = MemberMother.member()
            member.dept?.appointLeader(leader)

            val from = LocalDateTime.now().plusDays(1)
            val to = LocalDateTime.now()
            val vacationType = "vacationType"
            val vacationCreateRequestDto = VacationCreateRequestDto(from, to, "reason", vacationType)

            // when
            every { memberRepository.findByIdWithDeptAndLeader(0L) } returns member
            every {
                vacationRequestRepository.countInRangeFromBetweenToBy(
                    member.id!!,
                    vacationCreateRequestDto.from,
                    vacationCreateRequestDto.to
                )
            } returns 0L
            every {
                vacationRequestRepository.calculateRequestedVacationDays(from, to, vacationType)
            } returns 0.0
            every {
                vacationInfoRepository.findActualRemainingVacationDays(member.id!!, vacationType)
            } returns 1.0

            // then
            assertThatThrownBy {
                vacationService.requestVacation(member.id!!, vacationCreateRequestDto)
            }.isInstanceOf(RuntimeException::class.java)
        }
    }

    @Test
    fun `should save vacation when given valid member id and vacation create request dto`() {
        // given
        val member = MemberMother.member()
        val leader = MemberMother.member()
        member.dept?.appointLeader(leader)

        val from = LocalDateTime.now()
        val to = from.plusDays(2)
        val vacationCreateRequestDto = VacationCreateRequestDto(from, to, "reason", "vacationType")

        // when
        every { memberRepository.findByIdWithDeptAndLeader(member.id!!) } returns member
        every { vacationRequestRepository.countInRangeFromBetweenToBy(member.id!!, from, to) } returns 0L
        every {
            vacationRequestRepository.calculateRequestedVacationDays(
                from,
                to,
                vacationCreateRequestDto.vacationType
            )
        } returns Duration.between(from, to).toDays().toDouble()
        every {
            vacationInfoRepository.findActualRemainingVacationDays(
                member.id!!,
                vacationCreateRequestDto.vacationType
            )
        } returns 30.0
        every {
            codeRepository.findByGroupCodeAndCode("VACATION_TYPE", vacationCreateRequestDto.vacationType)
        } returns null

        // then
        assertThatThrownBy {
            vacationService.requestVacation(member.id!!, vacationCreateRequestDto)
        }.isInstanceOf(RuntimeException::class.java)
    }

    @Nested
    inner class GetVacationRequestList {
        @Test
        fun `should return empty result when find id page is empty`() {
            // given
            val member = MemberMother.member()

            // when
            every { memberRepository.findByIdOrNull(member.id!!) } returns member
            every {
                vacationRequestRepository.findIdsByRequesterIdPaging(any<Long>(), any<Pageable>())
            } returns Page.empty()

            // then
            val result = vacationService.getVacationRequestList(member.id!!, 0)
            assert(result.content.isEmpty())
            assert(result.pageNumber == 0)
        }
    }

    @Nested
    inner class UpdateVacationRequest {
    }
}